pkg/icingaredis/own_heartbeat.go

```diff
@@ -0,0 +1,92 @@
+package icingaredis
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/go-redis/redis/v8"
+	"github.com/icinga/icingadb/pkg/com"
+	"github.com/icinga/icingadb/pkg/logging"
+	"github.com/icinga/icingadb/pkg/utils"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"strconv"
+	"time"
+)
+
+const statSecs = 15 * 60
+
+// Stats of sync by kind for self-monitoring.
+var Stats = struct {
+	Config  *com.Stat `json:"config"`
+	State   *com.Stat `json:"state"`
+	History *com.Stat `json:"history"`
+	Overdue *com.Stat `json:"overdue"`
+}{
+	com.NewStat(statSecs),
+	com.NewStat(statSecs),
+	com.NewStat(statSecs),
+	com.NewStat(statSecs),
+}
+
+// haer abstracts HA to prevent package import cycles.
+type haer interface {
+	Responsible() bool
+}
+
+// OwnHeartbeat periodically writes heartbeats to a Redis stream.
+type OwnHeartbeat struct {
+	ctx    context.Context
+	client *Client
+	ha     haer
+	logger *logging.Logger
+}
+
+func NewOwnHeartbeat(ctx context.Context, client *Client, ha haer, logger *logging.Logger) *OwnHeartbeat {
+	oh := &OwnHeartbeat{ctx, client, ha, logger}
+
+	go oh.run()
+
+	return oh
+}
+
+func (oh *OwnHeartbeat) run() {
+	every3s := time.NewTicker(3 * time.Second)
+	defer every3s.Stop()
+
+	boolToIntStr := map[bool]string{false: "0", true: "1"}
+	wasResponsible := oh.ha.Responsible()
+	responsibleSince := time.Now()
+
+	for {
+		isResponsible := oh.ha.Responsible()
+		if isResponsible != wasResponsible {
+			wasResponsible = isResponsible
+			responsibleSince = time.Now()
+		}
+
+		statsJson, err := json.Marshal(&Stats)
+		if err != nil {
+			oh.logger.Errorw("Can't update own heartbeat", zap.Error(errors.WithStack(err)))
+		} else {
+			cmd := oh.client.XAdd(oh.ctx, &redis.XAddArgs{
+				Stream: "icingadb:heartbeat",
+				MaxLen: 1,
+				Values: []string{
+					"is-responsible", boolToIntStr[isResponsible],
+					"responsible-since", strconv.FormatInt(responsibleSince.UnixMilli(), 10),
+					"stats", string(statsJson),
```

Comment:
@julianbrost Just let me know if you want this gzipped (Go std/Boost).

Reply:
I'd leave it as is for now, having gzip-compressed data in Redis only makes things annoying. If we decide that this is too much data, I'd rather keep the buckets in Redis individually, so that you add and insert one value per second in Redis.